### ISSUE TITLE
[BUG] Amavis filling logs with errors, and emails don't seem to be sending nor receiving

### ISSUE THREAD

**Issue by ghnp5** [docker-mailserver/docker-mailserver#2666]
### Miscellaneous first checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.

### Affected Component(s)

Email delivery

### What happened and when does this occur?

Suddenly at around 4:30 AM today, this started happening, seems.

I got LOADS of errors like these:

> Jun 28 04:53:49 myserv amavis[881]: (!)_DIE: Suicide in child_init_hook: BDB can't connect db env. at /var/lib/amavis/db: BDB0087 DB_RUNRECOVERY: Fatal error, run database recovery, No such file or directory. at (eval 90) line 338.
> Jun 28 04:53:49 myserv amavis[882]: (!!)TROUBLE in child_init_hook: BDB can't connect db env. at /var/lib/amavis/db: BDB0087 DB_RUNRECOVERY: Fatal error, run database recovery, No such file or directory. at (eval 90) line 338.

What I did to "resolve" this issue was:

- Add `amavis: root` to `postfix-aliases.cf`
- Delete (rename) `/var/lib/amavis`
- Restart `mailserver`

Those errors seem to be gone now. But I've no idea why they started or what they mean.

Strange that this happened on the same day I upgraded to 11.1.0 (from 11.0.0)

After what I did above, I sent an email and attempted to receive an email.

Especially because I'm seeing a lot of these (even for the emails I sent and tried to receive):

> Jun 28 06:29:18 myserv postfix/error[2148]: 724A84809507: to=<email@destination.com>, relay=none, delay=0.77, delays=0.68/0/0/0.09, dsn=4.4.1, status=deferred (delivery temporarily suspended: connect to 127.0.0.1[127.0.0.1]:10024: Connection refused)

But after some time, seems the emails did process. Like if there was some queue holding up.

So, things seem OK now, but I'm confused about why did this happen.

### DMS version

v11.1.0

### What operating system is DMS running on?

Linux

### What instruction set architecture is DMS running on?

x86_64 / AMD64

### What container orchestration tool are you using?

Docker Compose

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [ ] I am inexperienced with mail servers
- [ ] I am uncomfortable with the CLI

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

**Comment by georglauterbach** [comment]
-- will review later today

**Comment by polarathene** [comment]
> Restart `mailserver`
> 
> But after some time, seems the emails did process. Like if there was some queue holding up.
> 
> So, things seem OK now, but I'm confused about why did this happen.

I think the mail queue will persist with the restart you did, all failed deliveries would be in that queue. So that seems fine.

Do check the [v11.1 release notes](https://github.com/docker-mailserver/docker-mailserver/releases/tag/v11.1.0), there was changes that are related to Amavis, as well as `/var/mail-state` (which may affect `/var/lib/amavis` too I think).

One of these changes is for the change detector service to `reload` Amavis service upon some config changes detected, this differs from how Postfix and Dovecot presently restart via `supervisorctl restart`, we've been considering switching those to `reload` as well instead of restarting them entirely.

---

When you say you restarted, how did you approach the restart? Did you do the same for when you updated the Docker image to v11.1? See [this discussions "summary" section of my answer](https://github.com/docker-mailserver/docker-mailserver/discussions/2652#discussioncomment-2983373), ideally you should "restart" with `docker-compose up -d --force-recreate`, and avoid `docker-compose restart` (or `stop` + `start`).

If it wasn't due to that it would be good to know how to reproduce, as it may very well be due to the change detector support with `reload` if unlucky.

**Comment by ghnp5** [comment]
Hey - thank you for your reply.

> When you say you restarted, how did you approach the restart?

I restarted through Portainer, which I assume is similar to `docker-compose stop / up -d`.

Although I'm not sure how a `recreate` would help, since `/var/mail-state` is mounted on the host as a volume.
Do you mean I should clear my `mail-state` completely?

If Amavis continues to behave since what I've done today, do you think I should be good and no further action is needed, or is there a risk that this may happen again, if I don't do something now?

Thank you very much!

**Comment by polarathene** [comment]
> I restarted through Portainer, which I assume is similar to `docker-compose stop / up -d`.

Ah ok, so that's probably the case, or it uses `docker-compose restart`.

Either way, those are not advised with our image, as the discussion link I shared explains, our startup scripts make quite a lot of assumptions when they adjust internal config files based on ENV settings, those changes usually assume the config is like it was in the release image, not that a change had already been made in the container before.

So if stuff goes wrong, it's often because of issues with that, the container should be destroyed/removed, so that the internal state is fresh from a new container. Any important state is persisted via volume mounts, which don't have such problems AFAIK.

> Although I'm not sure how a `recreate` would help, since `/var/mail-state` is mounted on the host as a volume.
> Do you mean I should clear my `mail-state` completely?

I don't recall any issues with that in the past. I was just making note that we changed some logic for when `/var/mail-state` is mounted, only enabled services sync internal data with the volume mount. I don't think it would cause a problem for upgrading the image, but just in case, since that amavis location is handled by it too, I thought I'd mention it.

The restart advice was for internal container state that you don't persist with a volume mount. If you went into the container and created a file in the container which isn't part of a volume mount, that file will still be there when you restart a container, sort of like suspending and resuming a laptop and your apps are still as they were. `docker-compose down` and `docker-compose up --force-recreate` both remove the old container, throwing away that internal state, so no old data lingers accidentally.

> If Amavis continues to behave since what I've done today, do you think I should be good and no further action is needed, or is there a risk that this may happen again, if I don't do something now?

No further action needed. If it happens again let us know. It's not clear why it broke like that, and we'd need a way to reproduce it before we could troubleshoot further.

It's possible that the problem only occurs when re-using the container_ (`restart`, or `stop` + `start`, or `up` twoice with no `down` in-between or no `--force-recreate` used_).

**Comment by ghnp5** [comment]
Thank you very much for your time replying.
I'll make sure that upgrading this image in the future will be done with `down` + `up -d --force-recreate`.

Although, when I upgrade, I always just run `up -d`, and the output of Docker Compose tells me "Recreating" / "Recreated" (when the image version changed), so I thought that'd do the job. Unless the output is misleading.
